src/shogun/machine/Machine.cpp

```diff
@@ -71,9 +71,38 @@ bool Machine::train(std::shared_ptr<Features> data)
 	return result;
 }
 
-bool Machine::train(const std::shared_ptr<Features>& data, const std::shared_ptr<Labels>& lab){
-	set_labels(lab);
-	return train(data);
+bool Machine::train(
+    const std::shared_ptr<Features>& data, const std::shared_ptr<Labels>& lab)
+{
+
```

Comment:
do you want to add the label encoder here already?

Reply:
I think it is good that add the label encoder in another PR, as this PR already has many things that need to be changed.